graphrag/index/workflows/v1/subflows/create_final_communities.py

```diff
@@ -0,0 +1,181 @@
+from pandas._typing import Suffixes
+from pandas.api.types import is_bool
+
+from graphrag.index.verbs.graph.unpack import unpack_graph_df
+from graphrag.index.verbs.overrides.aggregate import aggregate_df
+
+
+@verb(name="create_final_communities", treats_input_tables_as_immutable=True)
+def create_final_communities(
+    input: VerbInput,
+    callbacks: VerbCallbacks,
+    **_kwargs: dict,
+) -> VerbResult:
+    """All the steps to transform final communities."""
+    table = cast(pd.DataFrame, input.get_input())
+
+    graph_nodes = unpack_graph_df(table, callbacks, "clustered_graph", "nodes")
+    graph_edges = unpack_graph_df(table, callbacks, "clustered_graph", "edges")
+
+    source_clusters = graph_nodes.merge(
+        graph_edges,
+        left_on="label",
+        right_on="source",
+        how="inner",
+        suffixes=cast(Suffixes, ["_1", "_2"]),
+    )
+    target_clusters = graph_nodes.merge(
+        graph_edges,
+        left_on="label",
+        right_on="target",
+        how="inner",
+        suffixes=cast(Suffixes, ["_1", "_2"]),
+    )
+
+    concatenated_clusters = pd.concat(
+        [source_clusters, target_clusters], ignore_index=True
+    )
+
+    combined_clusters = _filter_clusters(concatenated_clusters)
+
+    cluster_relationships = aggregate_df(
+        combined_clusters,
+        aggregations=[
+            {
+                "column": "id_2",  # this is the id of the edge from the join steps above
+                "to": "relationship_ids",
+                "operation": "array_agg_distinct",
+            },
+            {
+                "column": "source_id_1",
+                "to": "text_unit_ids",
+                "operation": "array_agg_distinct",
+            },
+        ],
+        groupby=[
+            "cluster",
+            "level_1",  # level_1 is the left side of the join
+        ],
+    )
+
+    all_clusters = aggregate_df(
+        graph_nodes,
+        aggregations=[{"column": "cluster", "to": "id", "operation": "any"}],
+        groupby=["cluster", "level"],
+    )
+
+    joined = all_clusters.merge(
+        cluster_relationships,
+        left_on="id",
+        right_on="cluster",
+        how="inner",
+        suffixes=cast(Suffixes, ["_1", "_2"]),
+    )
+
+    filtered = _filter_secondary(joined)
+
+    titled = _create_title(filtered)
+
+    titled["raw_community"] = titled[
+        "id"
+    ]  # TODO: Rodrigo says "raw_community" is temporary
+    return create_verb_result(
+        cast(
+            Table,
+            titled[
+                [
+                    "id",
+                    "title",
+                    "level",
+                    "raw_community",
+                    "relationship_ids",
+                    "text_unit_ids",
+                ]
+            ],
+        )
+    )
+
+
+def _filter_clusters(table: pd.DataFrame) -> pd.DataFrame:
+    # level_1 is the left side of the join
+    # level_2 is the right side of the join
+    args = FilterArgs(
```

Comment:
So the filter is, where level_1 == level_2 ?

Reply:
Ah! I see, it is because of the suffixes haha. It confused me since level_1 and level_2 are also values.
Can we replace this verb invocation with a pandas one liner?

Reply:
Yes :)

Reply:
We could actually probably just get rid of the suffixes and accept the pandas defaults (x/y), which may get rid of the weird casting issue too